cloud/amazon/ec2_ami.py

```diff
@@ -267,17 +433,31 @@ def deregister_image(module, ec2):
     wait = module.params.get('wait')
     wait_timeout = int(module.params.get('wait_timeout'))
 
-    img = ec2.get_image(image_id)
-    if img == None:
-        module.fail_json(msg = "Image %s does not exist" % image_id, changed=False)
-
     try:
-        params = {'image_id': image_id,
-                  'delete_snapshot': delete_snapshot}
-
-        res = ec2.deregister_image(**params)
-    except boto.exception.BotoServerError, e:
-        module.fail_json(msg = "%s: %s" % (e.error_code, e.error_message))
+	img = ec2.get_image(image_id)
+    except boto.exception.EC2ResponseError, e:
+        if 'InvalidAMIID.NotFound' in e.error_code:
+    	    module.fail_json(msg = "Image %s does not exist" % image_id, changed=False)
+
+    # Get all snapshot ids associated with image
+    snapshots = []
+    if hasattr(img, 'block_device_mapping'):
+	for key in img.block_device_mapping:
+    	    snapshots.append(img.block_device_mapping[key].snapshot_id)
+    
+    # When trying to re-delete already deleted image it doesn't raise an exception
+    # It just returns an object without image attributes
+    if hasattr(img, 'id'):
+	try:
+    	    params = {'image_id': image_id,
+                      'delete_snapshot': delete_snapshot}
+
+    	    res = ec2.deregister_image(**params)
+
+	except boto.exception.BotoServerError, e:
+    	    module.fail_json(msg = "%s: %s" % (e.error_code, e.error_message))
+    else:
+	module.exit_json(msg = "Image %s has already been deleted" % image_id, changed=False)
```

Comment:
If image doesn't exist the exception is raised before this point, e.g. at 'InvalidAMIID.NotFound'